tensorly/decomposition/candecomp_parafac.py

```diff
@@ -1,66 +1,152 @@
+        scales = T.norm(factor, axis=0)
+        weights *= scales
+        scales_non_zero = T.where(scales==0, T.ones(T.shape(scales)), scales)
+        normalized_factors.append(factor/scales_non_zero)
+    return normalized_factors, weights
+
+
+def initialize_factors(tensor, rank, init='svd', random_state=None):
+    r"""Initialize factors used in `parafac`.
 
-        Computes a rank-`rank` decomposition of `tensor` [1]_ such that:
-        ``tensor = [| factors[0], ..., factors[-1] |]``
+    The type of initialization is set using `init`. If `init == 'random'` then
+    initialize factor matrices using `random_state`. If `init == 'svd'` then
+    initialize the `m`th factor matrix using the `rank` left singular vectors
+    of the `m`th unfolding of the input tensor.
 
     Parameters
     ----------
     tensor : ndarray
-    rank  : int
-            number of components
-    n_iter_max : int
-                 maximum number of iteration
+    rank : int
     init : {'svd', 'random'}, optional
-    tol : float, optional
-          tolerance: the algorithm stops when the variation in
-          the reconstruction error is less than the tolerance
-    random_state : {None, int, np.random.RandomState}
-    verbose : int, optional
-        level of verbosity
 
     Returns
     -------
     factors : ndarray list
-            list of factors of the CP decomposition
-            element `i` is of shape (tensor.shape[i], rank)
+        List of initialized factors of the CP decomposition where element `i`
+        is of shape (tensor.shape[i], rank)
 
-    References
-    ----------
-    .. [1] T.G.Kolda and B.W.Bader, "Tensor Decompositions and Applications",
-       SIAM REVIEW, vol. 51, n. 3, pp. 455-500, 2009.
     """
     rng = check_random_state(random_state)
 
     if init is 'random':
-        factors = [T.tensor(rng.random_sample((tensor.shape[i], rank)), **T.context(tensor)) for i in range(T.ndim(tensor))]
+        logger.info('[CPDecomp] Using random initialization')
+        factors = [T.tensor(rng.random_sample((tensor.shape[i], rank))) for i in range(T.ndim(tensor))]
+        factors, _ = normalize_factors(factors)
+        return factors
 
-    elif init is 'svd':
+    if init is 'svd':
+        logger.info('[CPDecomp] Using svd initialization')
         factors = []
         for mode in range(T.ndim(tensor)):
             U, _, _ = T.partial_svd(unfold(tensor, mode), n_eigenvecs=rank)
 
             if tensor.shape[mode] < rank:
                 # TODO: this is a hack but it seems to do the job for now
-                factor = T.tensor(np.zeros((U.shape[0], rank)), **T.context(tensor))
-                factor[:, tensor.shape[mode]:] = T.tensor(rng.random_sample((U.shape[0], rank - tensor.shape[mode])), **T.context(tensor))
+                factor = T.tensor(np.zeros((U.shape[0], rank)))
+                factor[:, tensor.shape[mode]:] = T.tensor(rng.random_sample((U.shape[0], rank - tensor.shape[mode])))
```

Comment:
Why did you remove the **T.context(tensor)? 
It is needed to preserve context (dtype and GPU/CPU)

Reply:
This was unintentional.